# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [4]:
def add_two(x):
    return x + 2

print(add_two(3))

# Lambda

lambda x: x +2
    
(lambda x: x +2)(3)

5


5

#### Saving to a Variable

In [6]:
f = lambda x,y: x+y+2
f(2,3)

7

#### Multiple Inputs

In [7]:
(lambda x,y,z:x*y*z)(3,5,8)

120

#### Passing a Lambda into a Function

In [9]:
def do_math(func, num):
    return func(num)

do_math(lambda x: x*x, 4)

16

#### Returning a Lambda from a Function

In [10]:
def return_fun():
    def multiply(num):
        return num*num
    
    return multiply

return_fun()(4)

16

In [11]:
def return_lamb(b,c):
    return lambda x,y: (b+c) - (x+y)

return_lamb(3,2)(4,3)

-2

#### If Statements within Lambdas

In [12]:
#Lambda x : True if (condition) else false
f = lambda num: num*2 if num>10 else num + 2

f(12)

24

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [14]:
f = lambda x: x**3

f(2)

8

In [15]:
#get good at reading Documentation because this is your lifelinealso
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [17]:
#Using Lambdas with sort to sort a list of dictionary by a key from highest to smallest
sorted([2, 8, 9], reverse=True)

[9, 8, 2]

In [19]:
w = {
    "name": "Wade",
    "age": 112
}

z = {
    "name": "Zak",
    "age": 13
}

people = [w,z]
print(people)

my_lambda = lambda person_dict:person_dict["age"]
print(sorted(people, key=my_lambda))

[{'name': 'Wade', 'age': 112}, {'name': 'Zak', 'age': 13}]
[{'name': 'Zak', 'age': 13}, {'name': 'Wade', 'age': 112}]


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [ ]:
# map(func,iterable(list,dict,tuple,set,etc...))
# Normally the usage of map happens with a pre-defined function -- but can use lambdas as well


In [21]:
def squared(num, num2):
    if num<10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2
numbers = [4, 10, 3]
more_nums = [4, 11, 20]

squared_nums_map = map(squared, numbers, more_nums)
print(list(squared_nums_map))

[(16, 16), (10, 11), (3, 20)]


#### Using Lambda's with Map

In [23]:
# map(Lambda x+2,list)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))
numbers = [4, 10, 3]
squared_nums = list(map((lambda x: x**2 if x<10 else x), numbers))
print(squared_nums)

[16, 10, 9]


In [24]:
squared_num2 = list(map((lambda x1, y1: (x1**2, y1**2) if x1 < 10 and y1<10 else (x1, y1)), numbers, more_nums))
squared_num2

[(16, 16), (10, 11), (3, 20)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [25]:
nums = [4, 2, 6, 8]

dub_minus_one = list(map(lambda num: (num*2)-1), nums)

print(dub_minus_one)

TypeError: map() must have at least two arguments.

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

#### Using Lambda's with Filter()

In [27]:
names = ["Kristina", "Alex", "Kyle"]

def k_names(name):
    if name[0].lower()

SyntaxError: invalid syntax (Temp/ipykernel_12612/1291511843.py, line 4)

In [29]:
names = ["Kristina", "Alex", "Kyle"]
    
new_names_lamb=list(filter(lambda name: True if name[0].lower()=='k' else False, names))
print(new_names_lamb)

['Kristina', 'Kyle']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [30]:
list = [2, 5, 19, 7, 8, 10]
from statistics import mean

mean(1,2,3)

below_mean = list(filter(lambda x:if x >= mean(nums) else False, nums))
print(mean(nums))
print(below_mean)

SyntaxError: invalid syntax (Temp/ipykernel_12612/1425558276.py, line 6)

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [ ]:
from functools import reduce
# reduce(function,iterable)



#### Using Lambda's with Reduce()

In [33]:
from functools import reduce

list_1 = [2, 4, 6, 8, 10]
def subtract_nums(num1, num2):
    return num1 - num2

result = reduce(subtract_nums, list_1)
result


-26

In [34]:
l_3 = [1, 1, 10, 2 ,30]

result_lamb = reduce(lambda num1, num2: num1-num2, l_3)
result_lamb

-42

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [39]:
l_5 = [1, 2, 3, 4]
multiply = reduce(lambda x,y: x*y, (l_5))
multiply

24

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [41]:
def recursive(i):
    print("hi")
    if i == 4:
        return
    else:
        i+=1
    recursive(i)
    
recursive(0)

hi
hi
hi
hi
hi


#### Implementing a Base Case

In [ ]:
def add_num(num):
    # set base case(when I want to stop)
    if num <= 1:
        print("add_nums(1)=1")
        return num
    else:
        print(f"add_nums{num})
        x = num - 1
        return num + add_nums(x)

#### Writing a Factorial Function

In [ ]:
# 5! = 5*4*3*2*1


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [42]:
n = 5
def subtract(num):
    if num <= 1:
        return 1
    else:
        return num - subtract(num-1)

subtract(n)

3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [44]:
list(range(5))

for i in range(5):
    print(i)

0
1
2
3
4


#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [46]:
def test():
    yield 1
    yield 2
    yield 3
    yield "a"
    yield ["a", "b"]
    
for i in test():
    print(i)

1
2
3
a
['a', 'b']


In [48]:
def my_range(n):
    for i in range(n):
        yield i
list(my_range(5))

[0, 1, 2, 3, 4]

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [50]:
def squared(stop, start, step=-1):
    while start>stop:
        yield start ** 2
        start -= step
        
for num in squared(0, 10, 1)

SyntaxError: invalid syntax (Temp/ipykernel_12612/2473482238.py, line 5)

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [51]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

new_places=list(filter(lambda name: True if name !=' ' and name != '' and name != '  ' else False, places))
print(new_places)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [52]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
sorted_author = sorted(author, key=lambda s: s.upper().split()[-1])
print(sorted_author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [1]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

cel_to_far = list(map(lambda x: (x[0], (9/5) * x[1] + 32), places))
print(cel_to_far)
    

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
